src/main/java/com/github/rschmitt/dynamicobject/internal/FressianSerialization.java

```diff
@@ -71,22 +72,30 @@ public static <T> T fromFressianByteArray(byte[] bytes) {
     }
 
     public static synchronized void registerType(Class type, String tag, ReadHandler readHandler, WriteHandler writeHandler) {
-        binaryTagCache.put(type, tag);
+        String currentTagForType = binaryTypeCache.put(type, tag);
```

Comment:
What's the thought process behind splitting this into its own cache? Should we disallow attempting registering a tag for 2 different classes?

Imagine someone calls `registerType(myClass1, "tag", ...)` then calls `registerTag(myClass2, "tag")`.

Reply:
> this split is more similar to EdnSerialization's implementation.

That's exactly it. We don't do similar validations/disallows for EdnSerialization, so not sure it is needed here. The earlier behavior had its own quirks. For example, registerTag did not remove from the cache, though registerType did. Open to ideas. I think the options are leave as is, or stick to one cache, with registerTag also removing the key from the cache now.

I also realized I can just use the binaryTagCache#removeTag call in place of the binaryTagCache#get call in the deregister methods.

Reply:
> Should we disallow attempting registering a tag for 2 different classes?

I think this is a good idea for a follow-up pull request.

> this split is more similar to EdnSerialization's implementation.

It's worth noting that Edn and Fressian have very different implementations (the former is based on read/print multi-methods in the Clojure runtime, the latter is based on a straightforward Java library), so it shouldn't _necessarily_ be surprising if the two classes have different approaches to state tracking. I haven't looked at this code in years, so I'd prefer to keep the changes in `EdnSerialization` and `FressianSerialization` as minimal as possible in order to facilitate a quick pull request. I can reassess the serialization implementation separately as part of some routine maintenance work, such as upgrading the build environment.

Reply:
I'd prefer reregistering a tag for a different type to be either all or nothing instead of this intermediate state of only updating some caches.

The original implementation was inconsistent in this way too due to the line where `fressianReadHandlers.putIfAbsent` is used , so I suppose you're not making the situation much worse.

Should we followup with a "last registration wins" change or "throw on existing registration of different type" change?